build-support/bin/native/bootstrap.sh

```diff
@@ -28,9 +28,10 @@ readonly CACHE_ROOT=${XDG_CACHE_HOME:-$HOME/.cache}/pants
 readonly CACHE_TARGET_DIR=${CACHE_ROOT}/bin/native-engine/${OS_ID}
 
 function calculate_current_hash() {
-  # Cached and unstaged files, with ignored files excluded.
+  # Cached and unstaged files, with ignored files excluded + $MODE mixed in.
   git ls-files -c -o --exclude-standard ${NATIVE_ROOT} | \
-    git hash-object -t blob --stdin-paths | fingerprint_data
+    git hash-object -t blob --stdin-paths | \
+      awk -v MODE="$MODE" '{print $1 "\n" MODE}' | fingerprint_data
```

Comment:
This needs more care, see [here](https://github.com/pantsbuild/pants/pull/4090/files/034f100eeb785c153315fb4f913ac3c98fe65641..d0ddf5cf2691540a2dc010d97714c07d0bcffea0#diff-b16e9df30403b637fa43f97251cfeb7bR102).

The MODE should probably not affect the version resource file that gets checked in, just the binary picked for use in dev mode. This may be too much to deal with for this RB - not coming up with simple ideas here.